src/xrpld/app/tx/detail/VaultDeposit.cpp

```diff
@@ -0,0 +1,171 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/VaultDeposit.h>
+
+#include <xrpld/app/misc/CredentialHelpers.h>
+#include <xrpld/app/tx/detail/MPTokenAuthorize.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/LedgerFormats.h>
+#include <xrpl/protocol/MPTIssue.h>
+#include <xrpl/protocol/STNumber.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+NotTEC
+VaultDeposit::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureSingleAssetVault))
+        return temDISABLED;
+
+    if (auto const ter = preflight1(ctx))
+        return ter;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    return preflight2(ctx);
+}
+
+TER
+VaultDeposit::preclaim(PreclaimContext const& ctx)
+{
+    auto const vault = ctx.view.read(keylet::vault(ctx.tx[sfVaultID]));
+    if (!vault)
+        return tecOBJECT_NOT_FOUND;
+
+    auto account = ctx.tx[sfAccount];
+    auto const assets = ctx.tx[sfAmount];
+    auto asset = vault->at(sfAsset);
+    if (assets.asset() != asset)
+        return tecWRONG_ASSET;
+
+    // Cannot deposit inside Vault an Asset frozen for the depositor
+    if (isFrozen(ctx.view, account, asset))
+        return tecFROZEN;
+
+    auto const share = MPTIssue(vault->at(sfMPTokenIssuanceID));
+    if (isFrozen(ctx.view, account, share))
+        return tecFROZEN;
+
+    if (vault->getFlags() == tfVaultPrivate && account != vault->at(sfOwner))
+    {
+        auto const err = requireAuth(
+            ctx.view, MPTIssue(vault->at(sfMPTokenIssuanceID)), account);
+        return err;
+
+        // The above will perform authorization check based on DomainID stored
+        // in MPTokenIssuance. Had this been a regular MPToken, it would also
+        // allow use of authorization granted by the issuer explicitly, but
+        // Vault does not have an MPT issuer (instead it uses pseudo-account).
+        //
+        // If we passed the above check then we also need to do similar check
+        // inside doApply(), in order to check for expired credentials.
+    }
+
+    return tesSUCCESS;
```

Comment:
There should be a check for `lsfMPTCanTransfer` if depositing MPT asset.

Reply:
Thanks, done